tools/approval-service/internal/approvalservice/eventprocessor/githubprocessor/githubhandler.go

```diff
@@ -0,0 +1,249 @@
+	GetAccessRequests(ctx context.Context, filter types.AccessRequestFilter) ([]types.AccessRequest, error)
+}
+
+var _ accessrequest.ReviewHandler = &Processor{}
+
+// Opt is a function that modifies the GitHubHandler.
+type Opt func(r *Processor) error
+
+// WithLogger sets the logger for the GitHubHandler.
+func WithLogger(log *slog.Logger) Opt {
+	return func(r *Processor) error {
+		r.log = log
+		return nil
+	}
+}
+
+var defaultOpts = []Opt{
+	WithLogger(slog.Default()),
+}
+
+// New creates a new Processor.
+func New(ctx context.Context, cfg config.GitHubSource, tele teleClient, store store.GitHubService, opts ...Opt) (*Processor, error) {
+	p := &Processor{
+		store:      store,
+		teleClient: tele,
+		log:        slog.Default(),
+		org:        cfg.Org,
+		repo:       cfg.Repo,
+		envs:       cfg.Environments,
+	}
+
+	for _, o := range append(defaultOpts, opts...) {
+		if err := o(p); err != nil {
+			return nil, fmt.Errorf("applying option: %w", err)
+		}
+	}
+
+	f, err := os.Open(cfg.Authentication.App.PrivateKeyPath)
+	if err != nil {
+		return nil, fmt.Errorf("opening private key: %w", err)
+	}
+	defer func() {
+		err = errors.Join(err, f.Close())
+	}()
+	pKey, err := io.ReadAll(f)
+	if err != nil {
+		return nil, fmt.Errorf("reading private key: %w", err)
+	}
+
+	client, err := github.NewForApp(ctx, cfg.Authentication.App.AppID, cfg.Authentication.App.InstallationID, pKey)
+	p.client = client
+
+	return p, nil
+}
+
+// FindExistingAccessRequest checks if an Access Request already exists for the given GitHub deployment review event.
+func (p *Processor) FindExistingAccessRequest(ctx context.Context, e githubevents.DeploymentReviewEvent) (types.AccessRequest, error) {
+	list, err := p.teleClient.GetAccessRequests(ctx, types.AccessRequestFilter{})
+	if err != nil {
+		return nil, fmt.Errorf("getting access requests: %w", err)
+	}
+	for _, req := range list {
+		labels := req.GetStaticLabels()
+		if labels == nil {
+			continue
+		}
+		if labels["organization"] == e.Organization && labels["repository"] == e.Repository && labels["environment"] == e.Environment && labels["workflow_run_id"] == strconv.Itoa(int(e.WorkflowID)) {
+			return req, nil
+		}
+	}
+
+	// No existing access request found.
+	return nil, nil
+}
+
+// CreateNewAccessRequest creates a new Access Request for the given GitHub deployment review event.
+func (p *Processor) CreateNewAccessRequest(ctx context.Context, e githubevents.DeploymentReviewEvent, req types.AccessRequest) (types.AccessRequest, error) {
+	// Get the workflow run info from GitHub.
+	runInfo, err := p.client.GeWorkflowRunInfo(ctx, e.Organization, e.Repository, e.WorkflowID)
+	if err != nil {
+		return nil, fmt.Errorf("getting workflow run info: %w", err)
+	}
+
+	reason, err := genReason(e, runInfo)
+	if err != nil {
+		return nil, fmt.Errorf("generating reason template: %w", err)
+	}
+	req.SetRequestReason(reason)
+	err = p.store.StoreWorkflowInfo(ctx, req, store.GitHubWorkflowInfo{
```

Comment:
is it correct to store the workflow info and access request before we are able to successfully create it and not deny it? Should this be moved down after the other parts?